iree/modules/hal/hal_module.c

```diff
@@ -301,19 +304,62 @@ IREE_VM_ABI_EXPORT(iree_hal_module_allocator_wrap_byte_buffer, riirii, r) {
         (offset + length - 1), buffer_length);
   }
 
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // Not all HAL implementations support wrapping buffers, and of those that do
+  // some may only support it in special situations such as when the buffer is
+  // not DEVICE_VISIBLE. The generated module can query whether the wrapping is
+  // possible and decide to use alternative means of upload if it is not; we
+  // make no policy (other than validity) over what's best here.
```

Comment:
FYI, I have this test code / TODO for some ways to test this optionally supported feature:

https://github.com/google/iree/blob/991f9cd423ba53ed0139fe33ab682c6ca55376c5/iree/hal/cts/buffer_mapping_test.cc#L33-L45

Reply:
what you have there should work - this just adds a way to query specifically for wrapping/importing memory that was not allocated from the allocator (a device may support mapping but not wrapping, like vulkan/cuda)